File: clase-04-07/TAD/List.cpp
#pragma once
#include "Iterable.cpp"

template <class T>
struct Node
{
  T data;
  Node *next = nullptr;
};

template <class T>
class ListIterator : public Iterator<T>
{
private:
  Node<T> *actual;

public:
  T next() override
  {
    T elem = actual->data;
    actual = actual->next;
    return elem;
  }

  bool hasNext() override
  {
    return actual != nullptr;
  }

  ListIterator(Node<T> *start)
  {
    actual = start;
  }
};

template <class T>
class List : public Iterable<T>
{
private:
  Node<T> *root = nullptr;

public:
  void insert(T elem)
  {
    Node<T> *newNode = new Node<T>;
    newNode->data = elem;
    newNode->next = root;
    root = newNode;
  }

  Iterator<T> *iterator()
  {
    return new ListIterator<T>(root);
  }
};

File: clase-04-07/TAD/List.java
package TAD;

import java.util.Iterator;

public class List<T> implements Iterable<T> {
  private Node root;

  private class Node {
    public T data;
    public Node next;
  }

  public void insert(T elem) {
    Node newNode = new Node();
    newNode.data = elem;
    newNode.next = root;
    root = newNode;
  }

  @Override
  public Iterator<T> iterator() {
    return new ListIterator(root);
  }

  private class ListIterator implements Iterator<T> {
    private Node actual;

    @Override
    public boolean hasNext() {
      return actual != null;
    }

    @Override
    public T next() {
      T elem = actual.data;
      actual = actual.next;
      return elem;
    }

    public ListIterator(Node start) {
      actual = start;
    }

  }

}
